File: java/src/Driver.java
/**
 * Created by zixuan on 2016/9/3.
 */

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;
import java.util.logging.FileHandler;
import java.util.logging.Level;

import net.sf.javaml.classification.Classifier;
import net.sf.javaml.classification.KNearestNeighbors;
import net.sf.javaml.core.Dataset;
import net.sf.javaml.core.DefaultDataset;
import net.sf.javaml.core.DenseInstance;
//import net.sf.javaml.core.SparseInstance;
//import net.sf.javaml.tools.ListTools;
//import weka.classifiers.trees.RandomForest;
//import weka.core.DenseInstance;
//import weka.core.Instances;


public class Driver {


    public Driver() throws Exception
    {
        /*
        if (!Init())
        {
            throw new Exception("Error when init");
        }
        */

        if (Config.Database.equals("SVC"))
            GetSVCSignatures();
        else if (Config.Database.equals(Config.SUSIGDatabase) || Config.Database.equals(Config.SelfDatabase))
            GetSignatures(Config.Database);
        else
            throw new Exception("DataSet not supported");

        // Utils.logger.log(Level.INFO, "Train set count: {0}", this.trainSets.size());
        // Utils.logger.log(Level.INFO, "Test set count: {0}", this.testSets.size());

        ArrayList<ArrayList<Double>> genuineX = new ArrayList<>();
        ArrayList<ArrayList<Double>> forgeryX = new ArrayList<>();
        for (int i = 0; i < this.trainSets.size(); ++i)
        {
            // Utils.logger.log(Level.INFO, "Training Process: {0}", i);
            PersonTrain personTrain = new PersonTrain(this.trainSets.get(i).GetGenuine(),
                    this.trainSets.get(i).GetForgery());
            ArrayList<ArrayList<Double>> genuine = new ArrayList<>();
            ArrayList<ArrayList<Double>> forgery = new ArrayList<>();
            personTrain.CalculateTrainSet(genuine, forgery);
            genuineX.addAll(genuine);
            forgeryX.addAll(forgery);
        }

        // Utils.logger.log(Level.INFO, "GenuineX count: {0}", genuineX.size());
        // Utils.logger.log(Level.INFO, "ForgeryX count: {0}", forgeryX.size());

        classifier = new KNearestNeighbors(5);
        Dataset dataset = new DefaultDataset();
        for (int i = 0; i < genuineX.size(); ++i)
        {
            double[] array = genuineX.get(i).stream().mapToDouble(d->d).toArray();
            DenseInstance instance = new DenseInstance(array, 1);
            dataset.add(instance);
        }
        for (int i = 0; i < forgeryX.size(); ++i)
        {
            double[] array = forgeryX.get(i).stream().mapToDouble(d->d).toArray();
            DenseInstance instance = new DenseInstance(array, 0);
            dataset.add(instance);
        }

        // Utils.logger.log(Level.INFO, "trainDataSet size: {0}", dataset.size());
        // Fit classifier
        classifier.buildClassifier(dataset);
    }

    public void test()
    {
        // Utils.logger.log(Level.INFO, "Start test");

        int genuineCount = 0;
        int forgeryCount = 0;
        int falseRejectCount = 0;
        int falseAcceptCount = 0;

        for (PersonSignatures personSignature : this.testSets)
        {
            PersonTest personTest = new PersonTest(new ArrayList<Signature>(personSignature.GetGenuine().subList(0, Config.RefCount)));
            ArrayList<Signature> genuineSet = personSignature.GetGenuine();
            ArrayList<Signature> forgerySet = personSignature.GetForgery();
            genuineCount += genuineSet.size();
            forgeryCount += forgerySet.size();
            for (int i = 0; i < genuineSet.size(); ++i)
            {
                ArrayList<Double> feature = personTest.CalcDis(genuineSet.get(i));
                double[] array = feature.stream().mapToDouble(d->d).toArray();
                DenseInstance instance = new DenseInstance(array, 1);
                Object predictedValue = this.classifier.classify(instance);
                if (!predictedValue.equals(instance.classValue())) {
                    // Utils.logger.log(Level.INFO, "FalseReject: {0} {1}", new Object[]{personSignature.GetUserName(), i});
                    falseRejectCount += 1;
                }
            }
            for (int i = 0; i < forgerySet.size(); ++i)
            {
                ArrayList<Double> feature = personTest.CalcDis(forgerySet.get(i));
                double[] array = feature.stream().mapToDouble(d->d).toArray();
                DenseInstance instance = new DenseInstance(array, 0);
                Object predictedValue = this.classifier.classify(instance);
                if (!predictedValue.equals(instance.classValue())) {
                    // Utils.logger.log(Level.INFO, "FalseAccepted: {0} {1}", new Object[]{personSignature.GetUserName(), i});
                    falseAcceptCount += 1;
                }
            }
        }

        Config.ShowConfig();
        double falseAcceptRate = 1 - falseRejectCount/(float)genuineCount;
        double falseRejectRate = 1 - falseAcceptCount/(float)forgeryCount;
        double avg = (falseAcceptRate + falseRejectRate) / 2;
        Utils.logger.log(Level.INFO, "Result : FalseAcceptRate {0} : FalseRejectRate {1} : Avg {2}", new Object[]{falseAcceptRate, falseRejectRate, avg});
    }

    private void GetSVCSignatures()
    {
        // Utils.logger.log(Level.INFO, "Getting signatures from svc database");
        ArrayList<PersonSignatures> totalSets = new ArrayList<>();
        for (int i = 1; i <= 40; ++i)
        {
            ArrayList<Signature> signatures = new ArrayList<>();
            for (int j = 1; j <= 40; ++j)
            {
                String fileName = "U"+i+"S"+j+".TXT";
                String filePath = Paths.get(Config.SvcDataPath, fileName).toString();
                int label = j <= 20 ? 1 : 0;
                try
                {
                    Signature signature = GetSignatureFromSvcFile(filePath, label);
                    signatures.add(signature);
                }
                catch (IOException e)
                {
                    // Utils.logger.log(Level.INFO, "Exception: {0}", e);
                }
            }
            PersonSignatures personSignatures = new PersonSignatures(String.valueOf(i),
                    new ArrayList<Signature>(signatures.subList(0, 20)),
                    new ArrayList<Signature>(signatures.subList(20, 40)));
            totalSets.add(personSignatures);
        }

        // Initialize trainSets and testSets
        this.trainSets = new ArrayList<>(totalSets.subList(0, (int) (totalSets.size() * Config.TrainSetRate)));
        this.testSets =  new ArrayList<>(totalSets.subList((int)(totalSets.size()*Config.TrainSetRate), totalSets.size()));
    }

    private Signature GetSignatureFromSvcFile(String filePath, int label) throws IOException
    {
        Path path = Paths.get(filePath);
        List<String> lines = Files.readAllLines(path);
        lines.remove(0);
        String userName = path.getFileName().toString().split("S")[0].split("U")[1];
        Signature signature = new Signature(userName, label);
        ArrayList<Double> X = new ArrayList<>();
        ArrayList<Double> Y = new ArrayList<>();
        ArrayList<Double> T = new ArrayList<>();
        ArrayList<Double> P = new ArrayList<>();
        for (String line : lines)
        {
            String items[] = line.split(" ");
            X.add(Double.valueOf(items[0]));
            Y.add(Double.valueOf(items[1]));
            T.add(Double.valueOf(items[2])/10.0);
            P.add(Double.valueOf(items[6]));
        }

        // Pre-process
        Processor.size_normalization(X, Y, Config.PreWidth, Config.PreHeight);
        Processor.location_normalization(X, Y);

        ArrayList<Double> VX = Processor.calculate_delta(X, T);
        ArrayList<Double> VY = Processor.calculate_delta(Y, T);

        signature.SetCom("X", X);
        signature.SetCom("Y", Y);
        signature.SetCom("T", T);
        signature.SetCom("P", P);
        signature.SetCom("VX", VX);
        signature.SetCom("VY", VY);
        return signature;
    }

    /*
    @Desc: Get signatures from data. The data will write to this.trainSets and this.testSets.
     */
    private void GetSignatures(String database) {
        // Utils.logger.log(Level.INFO, "Getting signatures from {0}", database);
        ArrayList<PersonSignatures> totalSets = new ArrayList<>();

        String forgeryFolderPath = new String();
        String genuineFolderPath = new String();
        if (database.equals(Config.SUSIGDatabase)) {
            forgeryFolderPath = Paths.get(Config.SusigDataPath, Config.VisualSubCorpus, Config.Forgery).toString();
            genuineFolderPath = Paths.get(Config.SusigDataPath, Config.VisualSubCorpus, Config.Genuine).toString();
        }
        else if (database.equals(Config.SelfDatabase))
        {
            forgeryFolderPath = Paths.get(Config.SelfDataPath, Config.Forgery).toString();
            genuineFolderPath = Paths.get(Config.SelfDataPath, Config.Genuine).toString();
        }
        HashMap<String, ArrayList<Signature>> userToForgery = GetUserToSigsMap(database, forgeryFolderPath, 0);
        HashMap<String, ArrayList<Signature>> userToGenuine = GetUserToSigsMap(database, genuineFolderPath, 1);
        Set<String> users = new HashSet<>(userToForgery.keySet());
        users.addAll(userToGenuine.keySet());

        for (String user : users)
        {
            PersonSignatures personSignature = new PersonSignatures(user, userToGenuine.get(user), userToForgery.get(user));
            totalSets.add(personSignature);
        }

        // Initialize trainSets and testSets
        this.trainSets = new ArrayList<>(totalSets.subList(0, (int) (totalSets.size() * Config.TrainSetRate)));
        this.testSets =  new ArrayList<>(totalSets.subList((int)(totalSets.size()*Config.TrainSetRate), totalSets.size()));
    }

    private HashMap<String, ArrayList<Signature>> GetUserToSigsMap(String database, String folderPath, int label)
    {
        // Utils.logger.log(Level.INFO, "Getting Signatures from {0}, label is {1}", new Object[]{folderPath, label});
        HashMap<String, ArrayList<Signature>> userToSignatures = new HashMap<>();
        File folder = new File(folderPath);
        File[] fileList = folder.listFiles();
        for (File file : fileList)
        {
            String userName = ExtractUserNameFromFile(database, file);
            if (!userToSignatures.containsKey(userName))
                userToSignatures.put(userName, new ArrayList<>());
            Signature signature = GetSignatureFromFile(database, file.getPath(), label);
            userToSignatures.get(userName).add(signature);
        }
        return userToSignatures;
    }

    private String ExtractUserNameFromFile(String database, File file)
    {
        String userName = userName = file.getName().split("_")[0];
        return userName;
    }

    private Signature GetSignatureFromFile(String database, String filePath, int label)
    {
        try {
            if (database.equals(Config.SUSIGDatabase)) {
                return GetSignatureFromSUSIGFile(filePath, label);
            }
            else if (database.equals(Config.SelfDatabase)) {
                return GetSignatureFromSelfFile(filePath, label);
            }
        } catch (Exception e) {
            // Utils.logger.log(Level.INFO, "Exception when get signature from {0} file. {1}", new Object[]{database, e.toString()});
        }
        return null;
    }

    private Signature GetSignatureFromSelfFile(String filePath, int label) throws IOException
    {
        Path path = Paths.get(filePath);
        List<String> lines = Files.readAllLines(path);
        String userName = path.getFileName().toString().split("_")[0];
        Signature signature = new Signature(userName, label);
        ArrayList<Double> T = new ArrayList<>();
        ArrayList<Double> X = new ArrayList<>();
        ArrayList<Double> Y = new ArrayList<>();
        ArrayList<Double> P = new ArrayList<>();
        for (String line : lines)
        {
            if (line.isEmpty())
                continue;
            String items[] = line.split(" ");
            T.add(Double.valueOf(items[0]));
            X.add(Double.valueOf(items[1]));
            Y.add(Double.valueOf(items[2]));
            P.add(Double.valueOf(items[3]));
        }

        Double maxY = Utils.GetMaxValue(Y);
        for (int i = 0; i < Y.size(); ++i)
            Y.set(i, maxY - Y.get(i));

        // Pre-process
        Processor.size_normalization(X, Y, Config.PreWidth, Config.PreHeight);
        Processor.location_normalization(X, Y);

        ArrayList<Double> VX = Processor.calculate_delta(X, T);
        ArrayList<Double> VY = Processor.calculate_delta(Y, T);

        signature.SetCom("X", X);
        signature.SetCom("Y", Y);
        signature.SetCom("T", T);
        signature.SetCom("P", P);
        signature.SetCom("VX", VX);
        signature.SetCom("VY", VY);
        return signature;
    }

    private Signature GetSignatureFromSUSIGFile(String filePath, int label) throws IOException
    {
        Path path = Paths.get(filePath);
        List<String> lines = Files.readAllLines(path);
        lines.remove(0);
        lines.remove(0);
        String userName = path.getFileName().toString().split("_")[0];
        Signature signature = new Signature(userName, label);
        ArrayList<Double> X = new ArrayList<>();
        ArrayList<Double> Y = new ArrayList<>();
        ArrayList<Double> T = new ArrayList<>();
        ArrayList<Double> P = new ArrayList<>();
        for (String line : lines)
        {
            String items[] = line.split(" ");
            X.add(Double.valueOf(items[0]));
            Y.add(Double.valueOf(items[1]));
            T.add(Double.valueOf(items[2]));
            P.add(Double.valueOf(items[3]));
        }

        // Pre-process
        Processor.size_normalization(X, Y, Config.PreWidth, Config.PreHeight);
        Processor.location_normalization(X, Y);

        ArrayList<Double> VX = Processor.calculate_delta(X, T);
        ArrayList<Double> VY = Processor.calculate_delta(Y, T);

        signature.SetCom("X", X);
        signature.SetCom("Y", Y);
        signature.SetCom("T", T);
        signature.SetCom("P", P);
        signature.SetCom("VX", VX);
        signature.SetCom("VY", VY);
        return signature;
    }

    private ArrayList<PersonSignatures> trainSets;

    private ArrayList<PersonSignatures> testSets;

    private Classifier classifier;

    private FileHandler fh;
}

File: python/driver.py
#coding=utf-8

import os
import cPickle

import settings
import processor
from person import PersonTraining, PersonTest

class BaseDriver(object):
    def __init__(self, config):
        self.config = config
        self.driver = self.config.model
        self.processor = processor.PreProcessor()
        if self.config.Dataset == settings.SUSIG:
            trainSets, testSets = self.get_signatures_from_susig()
        elif self.config.Dataset == settings.Self:
            trainSets, testSets = self.get_signatures_from_self()
        self.train_set = trainSets
        self.test_set = testSets

        self.config.logger.info("Train set count : %d" % len(trainSets))
        self.config.logger.info("Test set count : %d" % len(testSets))
        self.config.logger.info("Genuine count in train set : %d" % len(trainSets[0]["genuine"]))
        self.config.logger.info("Forgery count in train set : %d" % len(trainSets[0]["forgery"]))
        self.config.logger.info("Genuine count in test set : %d" % len(testSets[0]["genuine"]))
        self.config.logger.info("Forgery count in test set : %d" % len(testSets[0]["forgery"]))

    def pre_process_for_signle_signature(self, signature):
        T  = signature["T"]
        RX = signature["X"]
        RY = signature["Y"]
        if self.config.PreProcessorSwitch:
            if self.config.SizeNormSwitch:
                RX, RY = self.processor.size_normalization(RX, RY, 400, 200)
            if self.config.LocalNormalType== "mid":
                RX, RY = self.processor.location_normalization(RX, RY)
            elif self.config.LocalNormalType== "offset":
                RX, RY = self.processor.offset_to_origin_normalization(RX, RY)
            signature["X"] = RX
            signature["Y"] = RY
        return signature

    def reconstructSignature(self, signature):
        """
        Reconstruct signature to dictionary like object
        """
        def _calculate_delta(T, valueList):
            """
            This function just calculate the derivatives of valueList.
            """
            assert len(T) == len(valueList)
            newValueList = []
            for i in range(1, len(T)):
                slot = T[i] - T[i-1]
                value = (valueList[i] - valueList[i-1]) / slot \
                        if slot != 0 else None
                if value:
                    newValueList.append(value)
            return newValueList

        T = signature["T"]
        X = signature["X"]
        Y = signature["Y"]
        VX = _calculate_delta(T, X)
        VY = _calculate_delta(T, Y)
        signature["VX"] = VX
        signature["VY"] = VY
        return signature

    def get_signatures_from_susig(self):
        """
        Load original data from susig
        @return: trainSets, testSets
        """
        def _get_signatures_from_susig_folder(folder):
            
            signatures = {}
            for fileName in os.listdir(folder):
                filePath = os.path.join(folder, fileName)
                signature = {"T": [], "X": [], "Y": [], "P": []}
                with open(filePath) as fp:
                    lines = fp.readlines()
                    for line in lines[2:]:
                        line = line.strip()
                        if not line:
                            continue
                        items = line.split()
                        signature["X"].append(float(items[0]))
                        signature["Y"].append(float(items[1]))
                        signature["T"].append(float(items[2]))
                        signature["P"].append(float(items[3]))
                    personID = fileName.split(".")[0].split("_")[0]
                    if personID not in signatures:
                        signatures[personID] = []
                    signature = self.pre_process_for_signle_signature(signature)
                    signature = self.reconstructSignature(signature)
                    signatures[personID].append(signature)

            return signatures
        self.config.logger.info("Getting signatures from susig")
        trainSignatures = {}

        # For train forgery
        trainForgeryFolder = "../data/SUSig/VisualSubCorpus/FORGERY"
        trainForgerySubSigs = _get_signatures_from_susig_folder(trainForgeryFolder)
        for (personID, value) in trainForgerySubSigs.items():
            if personID not in trainSignatures:
                trainSignatures[personID] = {"genuine": [], "forgery": []}
            trainSignatures[personID]["forgery"].extend(value)

        # For train genuine session1
        trainGenuineSession1Folder = "../data/SUSig/VisualSubCorpus/GENUINE/SESSION1"
        trainGenuineSession1SubSigs= _get_signatures_from_susig_folder(trainGenuineSession1Folder)
        for (personID, value) in trainGenuineSession1SubSigs.items():
            if personID not in trainSignatures:
                trainSignatures[personID] = {"genuine": [], "forgery": []}
            trainSignatures[personID]["genuine"].extend(value)

        if self.config.MultiSession:
            # For train genuine session2
            trainGenuineSession2Folder = "../data/SUSig/VisualSubCorpus/GENUINE/SESSION2"
            trainGenuineSession2SubSigs = _get_signatures_from_susig_folder(trainGenuineSession2Folder)
            for (personID, value) in trainGenuineSession2SubSigs.items():
                if personID not in trainSignatures:
                    trainSignatures[personID] = {"genuine": [], "forgery": []}
                trainSignatures[personID]["genuine"].extend(value)

        trainSets = trainSignatures.values()

        # For test sets
        testSignatures = {}
        testForgeryFolder = "../data/SUSig/VisualSubCorpus/VALIDATION/VALIDATION_FORGERY"
        testForgerySubSigs = _get_signatures_from_susig_folder(testForgeryFolder)
        for (personID, value) in testForgerySubSigs.items():
            if personID not in testSignatures:
                testSignatures[personID] = {"genuine": [], "forgery": []}
            testSignatures[personID]["forgery"].extend(value)

        testGenuineFolder = "../data/SUSig/VisualSubCorpus/VALIDATION/VALIDATION_GENUINE"
        testGenuineSubSigs = _get_signatures_from_susig_folder(testGenuineFolder)
        for (personID, value) in testGenuineSubSigs.items():
            if personID not in testSignatures:
                testSignatures[personID] = {"genuine": [], "forgery": []}
            testSignatures[personID]["genuine"].extend(value)
        testSets = testSignatures.values()

        if self.config.Mode== settings.ModeTest:
            return testSets[:1], trainSets[:1]
        else:
            return testSets+trainSets[0:self.config.TrainSetCount-10], trainSets[self.config.TrainSetCount-10:]

    def get_signatures_from_self(self):
        """
            Load original data from self
            @return: trainSets, testSets
        """
        def _get_signatures_from_self_folder(folder):
            signatures = {}
            for fileName in os.listdir(folder):
                filePath = os.path.join(folder, fileName)
                signature = {"T": [], "X": [], "Y": [], "P": []}
                with open(filePath) as fp:
                    lines = fp.readlines()
                    for line in lines:
                        line = line.strip()
                        if not line:
                            continue
                        items = line.split()
                        signature["T"].append(float(items[0]))
                        signature["X"].append(float(items[1]))
                        signature["Y"].append(float(items[2]))
                        signature["P"].append(float(items[3]))
                    maxY = max(signature["Y"])
                    signature["Y"] = [maxY - y for y in signature["Y"]]
                    personID = fileName.split("_")[0]
                    if personID not in signatures:
                        signatures[personID] = []
                    signature = self.pre_process_for_signle_signature(signature)
                    signature = self.reconstructSignature(signature)
                    signatures[personID].append(signature)
            return signatures

        self.config.logger.info("Getting signatures from self")
        # For genuine signatures
        genuineSignatures = _get_signatures_from_self_folder(self.config.GenuineFolder)
        forgerySignatures = _get_signatures_from_self_folder(self.config.ForgeryFolder)
        
        totalSignatures = {}
        for (personID, value) in genuineSignatures.items():
            if personID not in totalSignatures:
                totalSignatures[personID] = {"genuine": [], "forgery": []}
            totalSignatures[personID]["genuine"].extend(value)
        for (personID, value) in forgerySignatures.items():
            if personID not in totalSignatures:
                totalSignatures[personID] = {"genuine": [], "forgery": []}
            totalSignatures[personID]["forgery"].extend(value)
        totalSet = totalSignatures.values()
        trainCount = int(len(totalSet) * self.config.TrainCountRate)
        if self.config.Mode == settings.ModeTest:
            return totalSet[:1], totalSet[1:2]
        else:
            return totalSet[:trainCount], totalSet[trainCount:]

    def test(self):
        pass

class ClassifyDriver(BaseDriver):
    def __init__(self, config):
        super(ClassifyDriver, self).__init__(config)

        genuineX = []
        forgeryX = []

        genuineY = []
        forgeryY = []

        # Training process
        for sigs in self.train_set:
            personTrain = PersonTraining(self.config, sigs)
            genuine, forgery = personTrain.calc_train_set()
            genuineX.extend(genuine)
            forgeryX.extend(forgery)

        genuineY = [1] * len(genuineX)
        forgeryY = [0] * len(forgeryX)

        trainX = genuineX + forgeryX
        trainY = genuineY + forgeryY

        self.driver.fit(trainX, trainY)
        with open(self.config.ModelDumpFilePath, "w") as fp:
            cPickle.dump(self.driver, fp)


    def test(self):
        self.config.logger.info("Start test")
        count = 1
        test_set = self.test_set
        if self.config.TrainSetInclude:
            test_set.extend(self.train_set)
        forgery_test_result = []
        genuine_test_result = []
        random_test_result = []

        genuine_test_dis = []
        forgery_test_dis = []

        for i in range(len(test_set)):
            one_test_set = test_set[i]
            self.config.logger.info("Test signature: %d" % count)
            count += 1
            personTest = PersonTest(self.config, one_test_set["genuine"][0:self.config.RefCount])
            genuine_set = one_test_set["genuine"][self.config.RefCount:]
            forgery_set = one_test_set["forgery"]
            random_set = []

            for j in range(len(genuine_set)):
                sig = genuine_set[j]
                dis = personTest.calc_dis(sig)
                res = self.driver.predict(dis)
                self.config.logger.info("Genuine Test: Result: %s, %s" % (res, dis))
                genuine_test_result.append(res)
                if (res != 1):
                    self.config.logger.fatal("FalseReject: uid: %d, sid: %d" % (i, j))

            for j in range(len(forgery_set)):
                sig = forgery_set[j]
                dis = personTest.calc_dis(sig)
                res = self.driver.predict(dis)
                self.config.logger.info("Forgery Test: Result: %s, %s" % (res, dis))
                forgery_test_result.append(res)
                if (res != 0):
                    self.config.logger.fatal("FalseAccept: uid: %d, sid: %d" % (i, j))

            if self.config.RandomForgeryInclude:
                for j in range(len(test_set)):
                    if i == j:
                        continue
                    random_set.extend(test_set[j]["genuine"])
                    random_set.extend(test_set[j]["forgery"])

                # train set included
                for one_train_set in self.train_set:
                    random_set.extend(one_train_set["genuine"])
                    random_set.extend(one_train_set["forgery"])

                for j in range(len(random_set)):
                    sig = random_set[j]
                    dis = personTest.calc_dis(sig)
                    res = self.driver.predict(dis)
                    self.config.logger.info("Random Test: Result: %s, %s" % (res, dis))
                    random_test_result.append(res)
                    if (res != 0):
                        self.config.logger.fatal("FalseAccept: uid: %d, sig: %d" % (i, j))

        self.config.logger.info("genuine test set count: %d" % len(genuine_test_result))
        self.config.logger.info("true accepted count: %d" % sum(genuine_test_result))
        self.config.logger.info("false rejected rate: %f" % (sum(genuine_test_result) / float(len(genuine_test_result))))

        self.config.logger.info("forgery test set count: %d" % len(forgery_test_result))
        self.config.logger.info("false accepted count: %d" % sum(forgery_test_result))
        self.config.logger.info("false accepted rate: %f" % (1 - sum(forgery_test_result) / float(len(forgery_test_result))))

        if self.config.RandomForgeryInclude:
            self.config.logger.info("random test set count: %d" % len(random_test_result))
            self.config.logger.info("false accepted count: %d" % sum(random_test_result))
            self.config.logger.info("false accepted rate: %f" % (1 - sum(random_test_result) / float(len(random_test_result))))


class RegressionDriver(BaseDriver):
    def __init__(self, config):
        super(RegressionDriver, self).__init__(config)
        genuineX = []
        forgeryX = []

        genuineY = []
        forgeryY = []

        # Training process
        for sigs in self.train_set:
            personTrain = PersonTraining(self.config, sigs)
            genuine, forgery = personTrain.calc_train_set()
            genuineX.extend(genuine)
            forgeryX.extend(forgery)

        # To adjust PCA result, 0 means genuine and 1 means forgery
        genuineY = [0.0] * len(genuineX)
        forgeryY = [1.0] * len(forgeryX)

        trainX = genuineX + forgeryX
        trainY = genuineY + forgeryY

        self.driver.fit(trainX, trainY)

    def test(self):
        self.config.logger.info("Start test")
        count = 1
        test_set = self.test_set
        if self.config.TrainSetInclude:
            test_set.extend(self.train_set)
        forgery_test_result = []
        genuine_test_result = []
        random_test_result = []

        genuine_test_dis = []
        forgery_test_dis = []

        falseRejectCount = 0
        falseAcceptSkillCount = 0
        falseAcceptRandomCount = 0

        for i in range(len(test_set)):
            one_test_set = test_set[i]
            self.config.logger.info("Test signature: %d" % count)
            count += 1
            personTest = PersonTest(self.config, one_test_set["genuine"][0:self.config.RefCount])
            genuine_set = one_test_set["genuine"][self.config.RefCount:]
            forgery_set = one_test_set["forgery"]
            random_set = []

            for j in range(len(genuine_set)):
                sig = genuine_set[j]
                dis = personTest.calc_dis(sig)
                if self.config.Regressor == "PCA":
                    res = self.driver.transform(dis)
                    res = res.tolist()[0][0]
                else:
                    res = self.driver.predict(dis)
                    res = res.tolist()[0]
                genuine_test_dis.append(res)
                self.config.logger.debug("Genuine Test: Result: %s, %s" % (res, dis))
                genuine_test_result.append(res)
                if (res > 0.5):
                    self.config.logger.debug("FalseReject: uid: %d, sid: %d" % (i, j))
                    falseRejectCount += 1

            for j in range(len(forgery_set)):
                sig = forgery_set[j]
                dis = personTest.calc_dis(sig)
                if self.config.Regressor == "PCA":
                    res = self.driver.transform(dis)
                    res = res.tolist()[0][0]
                else:
                    res = self.driver.predict(dis)
                    res = res.tolist()[0]
                forgery_test_dis.append(res)
                self.config.logger.debug("Forgery Test: Result: %s, %s" % (res, dis))
                forgery_test_result.append(res)
                if (res <= 0.5):
                    self.config.logger.debug("FalseAccept: uid: %d, sid: %d" % (i, j))
                    falseAcceptSkillCount += 1

            if self.config.RandomForgeryInclude:
                for j in range(len(test_set)):
                    if i == j:
                        continue
                    random_set.extend(test_set[j]["genuine"])
                    random_set.extend(test_set[j]["forgery"])

                # train set included
                for one_train_set in self.train_set:
                    random_set.extend(one_train_set["genuine"])
                    random_set.extend(one_train_set["forgery"])

                for j in range(len(random_set)):
                    sig = random_set[j]
                    dis = personTest.calc_dis(sig)
                    if self.config.Regressor == "PCA":
                        res = self.driver.transform(dis)
                        res = res.tolist()[0][0]
                    else:
                        res = self.driver.predict(dis)
                        res = res.tolist()[0]
                    forgery_test_dis.append(res)
                    self.config.logger.debug("Random Test: Result: %s, %s" % (res, dis))
                    random_test_result.append(res)
                    if (res <= 0.5):
                        self.config.logger.debug("FalseAccept: uid: %d, sig: %d" % (i, j))
                        falseAcceptRandomCount += 1

        self.config.logger.info("genuine test set count: %d" % len(genuine_test_result))
        self.config.logger.info("false reject count: %d" % falseRejectCount)
        self.config.logger.info("false rejected rate: %f" % (float(falseRejectCount) / float(len(genuine_test_result))))

        self.config.logger.info("forgery test set count: %d" % len(forgery_test_result))
        self.config.logger.info("false accepted count: %d" % falseAcceptSkillCount)
        self.config.logger.info("false accepted rate: %f" % (float(falseAcceptSkillCount) / float(len(forgery_test_result))))

        if self.config.RandomForgeryInclude:
            self.config.logger.info("random test set count: %d" % len(random_test_result))
            self.config.logger.info("false accepted count: %d" % falseAcceptRandomCount)
            self.config.logger.info("false accepted rate: %f" % (float(falseAcceptRandomCount) / float(len(random_test_result))))

        # Compute Equal Error Rate
        genuine_test_dis_list = sorted(genuine_test_dis, reverse=True) # desending order
        forgery_test_dis_list = sorted(forgery_test_dis) # asending order

        lastGap = 100.0
        genuineCount = len(genuine_test_dis_list)
        forgeryCount = len(forgery_test_dis_list)
        falseRejectRate = None
        falseAcceptRate = None
        for i in range(genuineCount):
            pivotal = genuine_test_dis_list[i]
            falseRejectRate = float(i) / genuineCount
            j = 0
            while j < forgeryCount and forgery_test_dis_list[j] <= pivotal:
                j += 1
            falseAcceptRate = float(j) / forgeryCount
            gap = abs(falseAcceptRate - falseRejectRate)
            if gap == 0.0:
                break;
            elif gap < lastGap:
                lastGap = gap
            else:
                break
        self.config.logger.info("falseRejectRate: %f, falseAcceptRate: %f, gap: %f" % (falseRejectRate, falseAcceptRate, lastGap))
        self.config.logger.info("TestResultGenuine : %s" % genuine_test_dis)
        self.config.logger.info("TestResultForgery : %s" % forgery_test_dis)
